ballerina/utils.bal

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2025 WSO2 LLC (http://www.wso2.com).
+//
+// WSO2 LLC. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/ai;
+
+# Converts metadata filters to Weaviate compatible filter format
+#
+# + filters - The metadata filters containing filter conditions and logical operators
+# + return - A map representing the converted filter structure or an error if conversion fails
+isolated function convertWeaviateFilters(ai:MetadataFilters filters) returns map<anydata>|ai:Error {
+    (ai:MetadataFilters|ai:MetadataFilter)[]? rawFilters = filters.filters;
+    if rawFilters == () || rawFilters.length() == 0 {
+        return {};
+    }
+    map<anydata>[] filterList = [];
+    foreach (ai:MetadataFilters|ai:MetadataFilter) filter in rawFilters {
+        if filter is ai:MetadataFilter {
+            map<anydata> filterMap = {};
+            string weaviateOp = check mapWeaviateOperator(filter.operator);
+            filterMap["path"] = [filter.key];
+            filterMap["operator"] = weaviateOp;
+            filterMap["valueText"] = filter.value;
+            filterList.push(filterMap);
+            continue;
+        }
+        map<anydata> nestedFilter = check convertWeaviateFilters(filter);
+        if nestedFilter.length() > 0 {
+            filterList.push(nestedFilter);
+        }
+    }
+    if filterList.length() == 0 {
+        return {};
+    }
+    if filterList.length() == 1 {
+        return filterList[0];
+    }
+    string weaviateCondition = check mapWeaviateCondition(filters.condition);
+    map<anydata> result = {

```

Comment:
Shall we create a Ballerina record for this.
anyway its not mandotory, since this is not a public API

Reply:
Since this method returns different fields in different scenarios, decided to keep the `map` type here. This is for keeping it more flexible rather than defining a record